### DISCUSSION TITLE
ImplicitRule is deprecated but there are no alternatives

### DISCUSSION THREAD

**Discussion by mnapoli** [laravel/framework#50798] [Q&A] [answered]
### Laravel Version

master, 10.x

### PHP Version

Any

### Description

Following https://github.com/laravel/framework/pull/19155, the `Rule` interface was deprecated in favor of `ValidationRule` (`Illuminate\Contracts\Validation` namespace).

`ImplicitRule` extends `Rule`, so it was deprecated as well (it cannot be changed to extend `ValidationRule` without a BC break).

The problem is that there is no alternative interface to `ImplicitRule`. That makes it impossible (AFAICT) to move away from the deprecated interfaces.

I'm not sure exactly _when_ the deprecated interfaces will be removed, it will be a blocker then.

I haven't looked at `InvokableRule` but it may be blocking for the same reasons.

### Steps To Reproduce

- implement `Illuminate\Contracts\Validation\ImplicitRule`
- try to move away from deprecated interfaces

**Comment by ahinkle** [comment, marked as answer]
There's not really a need as the new `ValidationRule` contract architecture is implicit. That's the alternative. 

https://laravel.com/docs/11.x/validation#using-rule-objects

**Reply by mnapoli** [reply]
Thank you for clarifying, the documentation doesn't mention that, nor any upgrade guide I could find 👍

**Reply by ahinkle** [reply]
Good point. Might be a good PR to the L11 upgrade docs. I'll review in a bit :)

**Reply by jelovac** [reply]
Implementing the `ValidationRule` contract in a custom validation rule does not automatically make the rule implicit by default, so this explanation is somewhat misleading. However, after reviewing the provided docs link, I created a dummy rule using the `php artisan make:rule Uppercase --implicit` flag and discovered that we need to include the `public $implicit = true;` property in our custom rules to make them implicit.

Am I missing something? Because I do not see how is this new solution better then having a dedicated contract or an attribute which we can put on a rule.

**Reply by fulopattila122** [reply]
I stumbled upon this today, and I discovered that in case you're using `Validator::extend(...)` it has a `Validator::extendImplicit(...)` counterpart which has the desired effect.

I know it's not 100% the same case as mentioned above, but it's tightly related.
